Sync time using public datetime API

zbrachinara opened the pull request:
Open for pull requests? Idk I'm the first one

The site can be fooled by an incorrect local system time, which I've heard has happened on some computers at the school. This approach takes the time from [worldtime api](https://worldtimeapi.org). or defaults to the system local time if the internet is not available (what seems to be the previous behavior).

zbrachinara committed:
fix not awaiting

zbrachinara committed:
function for getting current date accurately

zbrachinara commented:
I guess one important thing I haven't checked is if worldtimeapi.org is available from mcps proxy, so this probably is not a good idea to accept right away.

DefyGG commented:
I'm open to pull requests btw!

It's just that in this case, I prefer having it stay as the local computer time because when I test the program, I am able to check if the website will work for today, tmrw, etc... (by changing my computer date).

Maybe if I can somehow confirm that the website works for the rest of the year, I'll switch to using what you have instead.

Anyways, this is exactly the help I need. Thank you so much. If you have any other ideas that might help the website, let me know! I'll def check out your code!

zbrachinara commented on main.js:
```diff
@@ -90,8 +90,24 @@ window.dataLayer = window.dataLayer || [];
 
-  const calculateGoal = function() {
-    const date = new Date();
+  const reliableDate = async function() {
+    try {
+      return await dateEdt()
+    } catch {
+      console.log("catch activated, reverting to system time!")
+      return new Date()
+    }
+  }
```
For testing based on local time, I think this code would be perfect for your use! This statement switches between local and global time based on network availability, so with some modification it could be toggled manually.

DefyGG replied:
Nice! I'll check that out when I get home. Thanks for the feedback!

zbrachinara committed:
remove debug print and add documentation

zbrachinara committed:
Merge remote-tracking branch 'upstream/main'

zbrachinara committed:
create checkbox for toggling local time

zbrachinara committed:
react to dev options in js code

zbrachinara commented:
I've modified the `reliableDate` function to return the local time dependent on a boolean. That boolean is set on a checkbox that I've added to the html (not a frontend developer, so it looks kinda ugly 🤡).

DefyGG commented:
Haha dw, I have a slightly different idea for how I'm gonna do the local time testing. I'll try to merge your original request and make my edits.

zbrachinara committed:
Revert "create checkbox for toggling local time"

This reverts commit 06a06c810e6541548b4e24004aace7207ac3ed3a.

zbrachinara commented:
Got it. Is there any other action that I should take?

DefyGG commented:
Not for now. Actually, if you want that, I am working on a manual solution to the delayed schedules dilemma. If you want to work on an automatic version that requires no backend (using JS fetch command), that would be cool!

No worries though. The delayed schedule prediction is just an extra feature at this point.

DefyGG commented:
Sorry for taking so long till I merge your request, just working on a UI design change.

zbrachinara committed:
Revert "react to dev options in js code"

This reverts commit 15f740ba3f94ea6da77b2bd0c89ec3a97e17f978.

zbrachinara committed:
Merge remote-tracking branch 'upstream/main'

zbrachinara committed:
reintroduce utc query functions

zbrachinara commented:
After the major refactor, I realized how insanely slow this approach is (almost definitely because it requires doing three http queries back-to-back in each second). I'm going to go back and fix some stuff so that it doesn't end up making the entire website insanely slow

DefyGG commented:
Yep, I completely agree. I'm continuing to work on the UI update (sry its taking so long), but this update would be awesome!

zbrachinara committed:
switch to object-based date query

zbrachinara committed:
only request universal dates occasionally

zbrachinara committed:
block less on network time

zbrachinara committed:
fix anchor time not being set with network time

zbrachinara committed:
ensure that request runs once every interval

zbrachinara committed:
refactor locking code into Mutex class

zbrachinara commented:
These commits fix the large request volume, but it currently has issues automatically switching between global and local time. After that is fixed I'll move this PR out of draft status.

DefyGG commented:
That's awesome man! Just check if it merges nicely with the last refractor commit your friend made. I just finished the UI change. Now I'm just testing

DefyGG commented:
Also can I get your mcps email. I'm adding a credits section to the website

zbrachinara committed:
fix mutex bugs

zbrachinara committed:
remove unnecessary method from mutex

zbrachinara committed:
remove outdated mutex note on `DateProvider`
